gems/smithy/lib/smithy/views/client/protocol_plugin.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+module Smithy
+  module Views
+    module Client
+      # @api private
+      class ProtocolPlugin < View
+        def initialize(plan)
+          @plan = plan
+          @model = plan.model
+          super()
+        end
+
+        attr_reader :plan
+
+        def module_name
+          @plan.module_name
+        end
+
+        def protocol_name
+          raise ArgumentError, 'Invalid protocol' if (protocol = resolve_protocol).nil?
+
+          protocol
+        end
```

Comment:
TBD on exact Smithy plan of determining what protocol is "default" but Dowling is expecting that there is always a default protocol set.  (not sure if this is a REQUIREMENT since you can still build out a model w/o protocol)

Right now, I just look for the very first protocol that starts with `smithy.protocols` namespace

This gets me thinking about - should client generation happen even if there is no protocol set? Or should there be a raise?

Reply:
> Right now, I just look for the very first protocol that starts with smithy.protocols namespace

I think this is actually a problem we have in our model.  In the Java smithy codegen we could say, give me all protocol traits.  But in our model we don't have anything to tell us if a given trait is a protocol or not. (unless its a custom trait that we define ourselves).

We also need a way to register a Ruby protocol class to a protocol trait.  We could ptoentially solve both together by having welds "register" protocols - we would then ask all our welds for their protocols.  That gives us a list of all supported protocol traits and we can then match these up and find the first on the service and make that the default.